File: 769-min-cost-climbing-stairs/attempt_01_top_down.js
/**
 * @param {number[]} cost
 * @return {number}
 */
// top-down
 var minCostClimbingStairs = function(cost) {
    const dp = (i) => {
        // console.log('i: ', i);
        // handle base cases
        if (i < 2) return 0;
        // handle recursive case
        if (!memo[i]) {
            memo[i] = Math.min(dp(i - 1) + cost[i - 1], dp(i - 2) + cost[i - 2]);
        }
        // console.log('memo: ', memo);
        return memo[i];
    }
    const memo = {};
    return dp(cost.length);
    
};

File: 746-min-cost-climbing-stairs/attempt_01_top_down.py
from typing import List


class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        # define memo array
        memo = [None] * len(cost)
        # define base cases implicitly in memo
        memo[0], memo[1] = cost[0], cost[1]

        # define recursive function
        def dp(i):
            # handle recursive exploration
            if memo[i] is None:
                memo[i] = min(dp(i - 1), dp(i - 2)) + cost[i]
            return memo[i]
        dp(len(cost) - 1)
        return min(memo[-1], memo[-2])
